### ISSUE TITLE
实例分割分割结果受限于目标框

### ISSUE THREAD

**Issue by HePengguang** [ultralytics/ultralytics#14587]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

测试下来，由于目标框的存在，不管如果在后处理阶段调整目标框的范围，被分割目标的上下左右边界始终会被目标框切割，导致分割结果的边缘不平滑。请问能否提供一种训练方式，使分割效果不受限于bbox的回归，而仅仅训练mask。

**Comment by math-yyj** [comment]
路人回答：

那你可能需要的是solov2了；但是需要注意：solov2本身的结构太重

另一个思路是直接干掉yolov8的bbox回归，然后修改mask_loss的计算方式：所有目标都不再是只计算bbox范围内的loss，而是都计算整个feature_map上的loss；我没试过，不知道这效果会不会好

**Comment by pderrenger** [comment]
@math-yyj 感谢你的反馈。YOLOv8的实例分割确实依赖于目标框，这可能会导致分割结果的边缘不平滑。你提到的修改mask_loss计算方式的思路是一个有趣的方向，但需要进行大量实验验证其效果。目前，YOLOv8并不支持完全独立于目标框的分割训练。如果你有进一步的需求或建议，欢迎继续讨论。

**Comment by Burhan-Q** [comment]
Could be related to #11634 Maybe @Laughing-q could provide some input.

**Comment by HePengguang** [comment]
> 路人回答：
> 
> 那你可能需要的是solov2了；但是需要注意：solov2本身的结构太重
> 
> 另一个思路是直接干掉yolov8的bbox回归，然后修改mask_loss的计算方式：所有目标都不再是只计算bbox范围内的loss，而是都计算整个feature_map上的loss；我没试过，不知道这效果会不会好

谢谢您的建议啦！试了下solov2发现似乎还是会被box切割，也可能是我操作上的问题。再次感谢！

**Comment by HePengguang** [comment]
> @math-yyj 感谢你的反馈。YOLOv8的实例分割确实依赖于目标框，这可能会导致分割结果的边缘不平滑。你提到的修改mask_loss计算方式的思路是一个有趣的方向，但需要进行大量实验验证其效果。目前，YOLOv8并不支持完全独立于目标框的分割训练。如果你有进一步的需求或建议，欢迎继续讨论。

期待以后的更新！

**Comment by HePengguang** [comment]
> Could be related to #11634 Maybe @Laughing-q could provide some input.

谢谢您的建议！我立马去试试效果

**Comment by Liuzaiyang531** [comment]
请问解决了吗，我最近也遇到了这哥问题。导致边缘分割不精准。

**Comment by pderrenger** [comment]
目前，YOLOv8的实例分割确实依赖于目标框，这可能导致分割结果的边缘不平滑。我们建议尝试调整后处理阶段的参数，或者探索其他分割模型如SOLOv2。不过，这些模型可能会更重，影响性能。如果你有进一步的需求或建议，欢迎继续讨论。

**Comment by HePengguang** [comment]
> 请问解决了吗，我最近也遇到了这哥问题。导致边缘分割不精准。

没有，目前官方回复的几个解决方法暂时效果都不太好

**Comment by HePengguang** [comment]
> 目前，YOLOv8的实例分割确实依赖于目标框，这可能导致分割结果的边缘不平滑。我们建议尝试调整后处理阶段的参数，或者探索其他分割模型如SOLOv2。不过，这些模型可能会更重，影响性能。如果你有进一步的需求或建议，欢迎继续讨论。

好的好的，谢谢您的回复！

**Comment by Liuzaiyang531** [comment]
> > 请问解决了吗，我最近也遇到了这哥问题。导致边缘分割不精准。
> 
> 没有，目前官方回复的几个解决方法暂时效果都不太好

好吧，感谢您的回复！

**Comment by MingyunZhu** [comment]
To address this issue, I tried expanding the bounding boxes when generating them from the mask regions. The results are quite good (it only mitigates the issue rather than completely solving it), and it works well when the segmentation targets are not too close to each other.
I mainly modified the segments2boxes and segment2box functions in ultralytics\utils\ops.py during the training phase, expanding the top and bottom by 8% each, and the left and right by 4% each.
Consistent modifications are also required during the inference phase, involving the process_mask and process_mask_native methods. Basically, the bounding box needs to be expanded before cropping. I added a new expand_box method, and the code is as follows:


```
SEGMENT_BOX_EXPAND = 0.08  # 每边扩 10%，总宽高约 +20%
def segment2box(segment: np.ndarray, width: int = 640, height: int = 640) -> np.ndarray:
    x, y = segment.T
    if np.array([x.min() < 0, y.min() < 0, x.max() > width, y.max() > height]).sum() >= 3:
        x = x.clip(0, width)
        y = y.clip(0, height)
    inside = (x > 0) & (y > 0) & (x < width) & (y < height)
    x = x[inside]
    y = y[inside]
    if not any(x):
        return np.zeros(4, dtype=segment.dtype)

    x1, y1, x2, y2 = x.min(), y.min(), x.max(), y.max()
    bw, bh = x2 - x1, y2 - y1
    pad_x = bw * SEGMENT_BOX_EXPAND
    pad_y = bh * SEGMENT_BOX_EXPAND

    return np.array([
        max(0, x1-pad_x/2),
        max(0, y1 - pad_y),
        min(width, x2+pad_x/2),
        min(height, y2 + pad_y),
    ], dtype=segment.dtype)
```

```
def segments2boxes(segments):
    boxes = []
    for s in segments:
        x, y = s.T
        x1, y1, x2, y2 = x.min(), y.min(), x.max(), y.max()
        bw, bh = x2 - x1, y2 - y1
        pad_x = bw * SEGMENT_BOX_EXPAND
        pad_y = bh * SEGMENT_BOX_EXPAND
        boxes.append([
            max(0, x1-pad_x/2),
            max(0, y1 - pad_y),
            min(1, x2+pad_x/2),
            min(1, y2 + pad_y),
        ])
    return xyxy2xywh(np.array(boxes))
```

##################################
```
def expand_boxes(boxes: torch.Tensor, gain: float = 0.1, pad: float = 0.0) -> torch.Tensor:
    """Expand xyxy boxes by relative gain and/or absolute pad (same unit as boxes)."""
    boxes = boxes.clone()
    w = boxes[:, 2] - boxes[:, 0]
    h = boxes[:, 3] - boxes[:, 1]
    boxes[:, 0] -= w*gain/2 +pad
    boxes[:, 1] -= h * gain + pad
    boxes[:, 2] += w*gain/2 + pad
    boxes[:, 3] += h * gain + pad
    return boxes
```


```
def process_mask(protos, masks_in, bboxes, shape, upsample: bool = False):
    """Apply masks to bounding boxes using mask head output.

    Args:
        protos (torch.Tensor): Mask prototypes with shape (mask_dim, mask_h, mask_w).
        masks_in (torch.Tensor): Mask coefficients with shape (N, mask_dim) where N is number of masks after NMS.
        bboxes (torch.Tensor): Bounding boxes with shape (N, 4) where N is number of masks after NMS.
        shape (tuple): Input image size as (height, width).
        upsample (bool): Whether to upsample masks to original image size.

    Returns:
        (torch.Tensor): A binary mask tensor of shape [n, h, w], where n is the number of masks after NMS, and h and w
            are the height and width of the input image. The mask is applied to the bounding boxes.
    """
    c, mh, mw = protos.shape  # CHW
    masks = (masks_in @ protos.float().view(c, -1)).view(-1, mh, mw)  # NHW

    width_ratio = mw / shape[1]
    height_ratio = mh / shape[0]
    ratios = torch.tensor([[width_ratio, height_ratio, width_ratio, height_ratio]], device=bboxes.device)

    boxes = expand_boxes(bboxes, gain=0.08) * ratios
    masks = crop_mask(masks, boxes=boxes)  # NHW
    # masks = crop_mask(masks, boxes=bboxes * ratios)  # NHW
    if upsample:
        masks = F.interpolate(masks[None], shape, mode="bilinear")[0]  # NHW
    return masks.gt_(0.0).byte()
```


```
def process_mask_native(protos, masks_in, bboxes, shape):
    """Apply masks to bounding boxes using mask head output with native upsampling.

    Args:
        protos (torch.Tensor): Mask prototypes with shape (mask_dim, mask_h, mask_w).
        masks_in (torch.Tensor): Mask coefficients with shape (N, mask_dim) where N is number of masks after NMS.
        bboxes (torch.Tensor): Bounding boxes with shape (N, 4) where N is number of masks after NMS.
        shape (tuple): Input image size as (height, width).

    Returns:
        (torch.Tensor): Binary mask tensor with shape (N, H, W).
    """
    c, mh, mw = protos.shape  # CHW
    masks = (masks_in @ protos.float().view(c, -1)).view(-1, mh, mw)
    masks = scale_masks(masks[None], shape)[0]  # NHW
    # masks = crop_mask(masks, bboxes)  # NHW
    masks = crop_mask(masks, expand_boxes(bboxes, gain=0.08))  # NHW
    return masks.gt_(0.0).byte()
```
